plasmapy/formulary/relativity.py

```diff
@@ -150,3 +156,354 @@ def relativistic_energy(m: u.kg, v: u.m / u.s) -> u.Joule:
+        self,
+        velocity_like_arguments: Dict[str, Union[u.Quantity, Real]],
+    ):
+
+        not_none_arguments = {
+            key: value
+            for key, value in velocity_like_arguments.items()
+            if value is not None
+        }
+
+        if len(not_none_arguments) > 1:
+            raise ValueError(
+                "RelativisticBody can accept no more than one of the "
+                "following arguments: V, v_over_c, momentum, total_energy, "
+                "kinetic_energy, and lorentz_factor."
+            )
+
+        if not not_none_arguments:
+            return {"velocity": np.nan * u.m / u.s}
+
+        return not_none_arguments
+
+    def _store_velocity_like_argument(
+        self, speed_like_input: Dict[str, Union[u.Quantity, Real]]
+    ):
+        """
+        Take the velocity-like argument and store it via the setter for
+        the corresponding attribute.
+        """
+        name = list(speed_like_input.keys())[0]
+        value = speed_like_input[name]
+        if self._dtype:
+            value = u.Quantity(value, dtype=self._dtype)
+        setattr(self, name, value)
+
+    @validate_quantities(
+        V={"can_be_inf": False, "none_shall_pass": True},
+        momentum={"can_be_inf": False, "none_shall_pass": True},
+        total_energy={"can_be_negative": False, "none_shall_pass": True},
+        kinetic_energy={"can_be_negative": False, "none_shall_pass": True},
+    )
+    def __init__(
+        self,
+        particle: Union[ParticleLike, u.Quantity],
+        V: u.m / u.s = None,
+        momentum: u.kg * u.m / u.s = None,
+        *,
+        total_energy: u.J = None,
+        kinetic_energy: u.J = None,
+        v_over_c: Optional[Real] = None,
+        lorentz_factor: Optional[Real] = None,
+        Z: Optional[Integral] = None,
+        mass_numb: Optional[Integral] = None,
+        dtype: Optional[DTypeLike] = np.longdouble,
+    ):
+
+        self._particle = _physical_particle_factory(particle, Z=Z, mass_numb=mass_numb)
+        self._dtype = dtype
+
+        velocity_like_inputs = {
+            "velocity": V,
+            "momentum": momentum,
+            "total_energy": total_energy,
+            "kinetic_energy": kinetic_energy,
+            "v_over_c": v_over_c,
+            "lorentz_factor": lorentz_factor,
+        }
+
+        speed_like_input = self._get_speed_like_input(velocity_like_inputs)
+        self._store_velocity_like_argument(speed_like_input)
+
+    def __repr__(self):
+        return f"RelativisticBody({self.particle}, {self.velocity})"
+
+    @property
+    def particle(self) -> Union[CustomParticle, Particle, ParticleList]:
+        """
+        Representation of the particle(s).
+
+        Returns
+        -------
+        |Particle|, |CustomParticle|, or |ParticleList|
+        """
+        return self._particle
+
+    @property
+    @validate_quantities
+    def mass(self) -> u.kg:
+        r"""
+        The rest mass of the body, :math:`m_0`\ .
```

Comment:
Maybe this should, then, be renamed `rest_mass`? `mass_at_rest`, so it pops out upon a a `body.m<tab>`?

Reply:
This was something I was wondering about too, but decided to keep the attribute name of `mass` (at least for now) in case we wanted to duck type a `RelativisticBody` as a `Particle`.